File: app/src/main/java/de/westnordost/streetcomplete/location/LocationRequestFragment.java
package de.westnordost.streetcomplete.location;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AlertDialog;

import de.westnordost.streetcomplete.R;


/** Manages the process to ensure that the app can access the user's location. Two steps:
 *  <ol>
 *      <li>ask for permission</li>
 *      <li>ask for location to be turned on</li>
 *  </ol>
 *
 *  This fragment reports back to the Activity it is attached to via LocationRequestListener.
 *  The process is started via {@link #startRequest()} */
public class LocationRequestFragment extends Fragment
{
	public static final String
			ACTION_FINISHED = "de.westnordost.LocationRequestFragment.FINISHED",
			STATE = "state";

	private static final int LOCATION_PERMISSION_REQUEST = 1;
	private static final int LOCATION_TURN_ON_REQUEST = 2;

	private LocationState state;
	private boolean inProgress;
	private BroadcastReceiver locationProviderChangedReceiver;

	public LocationRequestFragment()
	{
		super();
		state = null;
	}

	/** Start location request process. When already started, will not be started again. */
	public void startRequest()
	{
		if(!inProgress)
		{
			inProgress = true;
			state = null;
			nextStep();
		}
	}

	private void nextStep()
	{
		if(state == null || state == LocationState.DENIED)
		{
			requestLocationPermissions();
		}
		else if(state == LocationState.ALLOWED)
		{
			requestLocationSettingsToBeOn();
		}
		else if(state == LocationState.ENABLED)
		{
			finish();
		}
	}

	private void finish()
	{
		inProgress = false;
		Intent intent = new Intent(ACTION_FINISHED);
		intent.putExtra(STATE, state.name());
		LocalBroadcastManager.getInstance(getContext()).sendBroadcast(intent);
	}

	@Override public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
										   @NonNull int[] grantResults)
	{
		// must be for someone else...
		if(requestCode != LOCATION_PERMISSION_REQUEST) return;
		if(permissions.length == 0 || !permissions[0].equals(Manifest.permission.ACCESS_FINE_LOCATION))
			return;

		if(grantResults[0] == PackageManager.PERMISSION_GRANTED)
		{
			requestLocationPermissions(); // retry then...
		}
		else
		{
			new AlertDialog.Builder(getContext())
					.setMessage(R.string.no_location_permission_warning)
					.setPositiveButton(R.string.retry,	(dialog, which) -> requestLocationPermissions())
					.setNegativeButton(android.R.string.cancel, (dialog, which) -> deniedlocationPermissions())
					.setOnCancelListener(dialog -> deniedlocationPermissions())
					.show();
		}
	}

	@Override public void onActivityResult(int requestCode, int resultCode, Intent data)
	{
		// must be for someone else...
		if(requestCode != LOCATION_TURN_ON_REQUEST) return;
		// we ignore the resultCode, because we always get Activity.RESULT_CANCELED. Instead, we
		// check if the conditions are fulfilled now
		requestLocationSettingsToBeOn();
	}

	private void deniedlocationPermissions()
	{
		state = LocationState.DENIED;
		finish();
	}

	private void requestLocationPermissions()
	{
		if (LocationUtil.hasLocationPermission(getContext()))
		{
			state = LocationState.ALLOWED;
			nextStep();

		} else {
			requestPermissions(new String[]{ Manifest.permission.ACCESS_FINE_LOCATION },
					LOCATION_PERMISSION_REQUEST);
		}
	}

	private void requestLocationSettingsToBeOn()
	{
		if(LocationUtil.isLocationOn(getContext()))
		{
			state = LocationState.ENABLED;
			nextStep();
		}
		else
		{
			final AlertDialog dlg = new AlertDialog.Builder(getContext())
					.setMessage(R.string.turn_on_location_request)
					.setPositiveButton(android.R.string.yes, (dialog, which) ->
					{
						dialog.dismiss();
						Intent viewIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
						startActivityForResult(viewIntent, LOCATION_TURN_ON_REQUEST);
					})
					.setNegativeButton(android.R.string.no,	(dialog, which) -> cancelTurnLocationOnDialog())
					.setOnCancelListener(dialog -> cancelTurnLocationOnDialog())
					.create();

			// the user may turn on location in the pull-down-overlay, without actually going into
			// settings dialog
			registerForLocationProviderChanges(dlg);

			dlg.show();
		}
	}

	private void cancelTurnLocationOnDialog()
	{
		unregisterForLocationProviderChanges();
		finish();
	}

	private void registerForLocationProviderChanges(final AlertDialog dlg)
	{
		locationProviderChangedReceiver = new BroadcastReceiver()
		{
			@Override public void onReceive(Context context, Intent intent)
			{
				dlg.dismiss();
				unregisterForLocationProviderChanges();
				requestLocationSettingsToBeOn();
			}
		};

		getActivity().registerReceiver(locationProviderChangedReceiver, LocationUtil.createLocationAvailabilityIntentFilter());
	}

	private void unregisterForLocationProviderChanges()
	{
		if(locationProviderChangedReceiver != null)
		{
			getActivity().unregisterReceiver(locationProviderChangedReceiver);
			locationProviderChangedReceiver = null;
		}
	}

	@Override public void onStop()
	{
		super.onStop();
		unregisterForLocationProviderChanges();
	}

	public LocationState getState()
	{
		return state != null ? state : LocationState.DENIED;
	}

	@Override public void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		if(state != null) outState.putString("locationState", state.name());
		outState.putBoolean("inProgress", inProgress);
	}

	@Override public void onActivityCreated(@Nullable Bundle savedInstanceState)
	{
		super.onActivityCreated(savedInstanceState);
		if(savedInstanceState != null)
		{
			String stateName = savedInstanceState.getString("locationState");
			if(stateName != null) state = LocationState.valueOf(stateName);
			inProgress = savedInstanceState.getBoolean("inProgress");
		}
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/location/LocationRequestFragment.kt
package de.westnordost.streetcomplete.location

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import de.westnordost.streetcomplete.R

/** Manages the process to ensure that the app can access the user's location. Two steps:
 *
 *  1. ask for permission
 *  2. ask for location to be turned on
 *
 * This fragment reports back via a local broadcast with the intent LocationRequestFragment.ACTION_FINISHED
 * The process is started via [.startRequest]  */
class LocationRequestFragment : Fragment() {

    var state: LocationState? = null
    private set

    private var inProgress = false
    private var locationProviderChangedReceiver: BroadcastReceiver? = null

    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission(), ::onRequestLocationPermissionsResult
    )
    private val startActivityForResultLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult(), ::onRequestLocationSettingsToBeOnResult
    )

    /* Lifecycle */

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) {
            val stateName = savedInstanceState.getString("locationState")
            if (stateName != null) state = LocationState.valueOf(stateName)
            inProgress = savedInstanceState.getBoolean("inProgress")
        }
    }


    override fun onStop() {
        super.onStop()
        unregisterForLocationProviderChanges()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (state != null) outState.putString("locationState", state!!.name)
        outState.putBoolean("inProgress", inProgress)
    }

    /** Start location request process. When already started, will not be started again.  */
    fun startRequest() {
        if (!inProgress) {
            inProgress = true
            state = null
            nextStep()
        }
    }

    private fun nextStep() {
        if (state == null || state == LocationState.DENIED) {
            checkLocationPermissions()
        } else if (state == LocationState.ALLOWED) {
            checkLocationSettings()
        } else if (state == LocationState.ENABLED) {
            finish()
        }
    }

    private fun finish() {
        inProgress = false
        val intent = Intent(ACTION_FINISHED)
        intent.putExtra(STATE, state!!.name)
        LocalBroadcastManager.getInstance(requireContext()).sendBroadcast(intent)
    }

    /* Step 1: Ask for permission */

    private fun checkLocationPermissions() {
        if (requireContext().hasLocationPermission) {
            state = LocationState.ALLOWED
            nextStep()
        } else {
            requestLocationPermissions()
        }
    }

    private fun requestLocationPermissions() {
        val activity = requireActivity()
        if (ActivityCompat.shouldShowRequestPermissionRationale(activity, ACCESS_FINE_LOCATION)) {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.no_location_permission_warning_title)
                .setMessage(R.string.no_location_permission_warning)
                .setPositiveButton(android.R.string.ok) { _, _ -> requestPermissionLauncher.launch(ACCESS_FINE_LOCATION) }
                .setNegativeButton(android.R.string.cancel) { _, _ -> deniedLocationPermissions() }
                .setOnCancelListener { deniedLocationPermissions() }
                .show()
        } else {
            requestPermissionLauncher.launch(ACCESS_FINE_LOCATION)
        }
    }

    private fun onRequestLocationPermissionsResult(isGranted: Boolean) {
        if (isGranted) checkLocationPermissions()
        else deniedLocationPermissions()
    }

    private fun deniedLocationPermissions() {
        state = LocationState.DENIED
        finish()
    }

    /* Step 1: Ask for location to be turned on */

    private fun checkLocationSettings() {
        if (requireContext().isLocationEnabled) {
            state = LocationState.ENABLED
            nextStep()
        } else {
            requestLocationSettingsToBeOn()
        }
    }

    private fun requestLocationSettingsToBeOn() {
        val dlg = AlertDialog.Builder(requireContext())
            .setMessage(R.string.turn_on_location_request)
            .setPositiveButton(android.R.string.ok) { dialog, _ ->
                dialog.dismiss()
                startActivityForResultLauncher.launch(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS))
            }
            .setNegativeButton(android.R.string.cancel) { _, _ -> cancelTurnLocationOnDialog() }
            .setOnCancelListener { cancelTurnLocationOnDialog() }
            .create()

        // the user may turn on location in the pull-down-overlay, without actually going into
        // settings dialog
        registerForLocationProviderChanges(dlg)
        dlg.show()
    }

    private fun onRequestLocationSettingsToBeOnResult(result: ActivityResult) {
        // we ignore the resultCode, because we always get Activity.RESULT_CANCELED. Instead, we
        // check if the conditions are fulfilled now
        checkLocationSettings()
    }

    private fun registerForLocationProviderChanges(dlg: AlertDialog) {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                dlg.dismiss()
                unregisterForLocationProviderChanges()
                checkLocationSettings()
            }
        }
        requireContext().registerReceiver(receiver, createLocationAvailabilityIntentFilter())
        locationProviderChangedReceiver = receiver
    }

    private fun unregisterForLocationProviderChanges() {
        locationProviderChangedReceiver?.let { receiver ->
            requireContext().unregisterReceiver(receiver)
        }
        locationProviderChangedReceiver = null
    }

    private fun cancelTurnLocationOnDialog() {
        unregisterForLocationProviderChanges()
        finish()
    }

    companion object {
        const val ACTION_FINISHED = "de.westnordost.LocationRequestFragment.FINISHED"
        const val STATE = "state"
    }
}
